[pcache] Cache and AsyncCache implementations

nmacchioni opened the pull request:
Summary:
Implemented caching abstractions: `Cache` and `AsyncCache`.

`Cache` provides an abstraction for defining simple key -> value stores with get and put functionality. We propose using `Cache` for implementations with very low (microseconds) overhead, for example an in-memory cache.

`AsyncCache` provides an abstraction for defining simple key -> value stores with asynchronous get and put functionality. We propose using `AsyncCache` for implementations with medium to high (> millisecond) overhead, for example an on-disk cache.

We provide an initial extension of `Cache` in the form of `InMemoryCache`. `InMemoryCache` provides fast, in-memory caching that can be later used to memoize more expensive cache accesses. `InMemoryCache` also provides a custom constructor `InMemoryCache.from_env_var` that can be used to pre-populate the in-memory cache, which will be helpful for enabling determinism in the future.

We also provides extensions of `AsyncCache`. `OnDiskCache` subclasses `AsyncCache` and serves as a generic on-disk caching implementation with atomic, write-once guarantees. `OnDiskCache` is semi-generic, allowing subclassing to alter the output directory. `InductorOnDiskCache` subclasses `OnDiskCache` to create an Inductor-specific on-disk cache that outputs to Inductor's default caching directory.

Test Plan:
`Cache` Tests:
1. Get -> Set -> Get
- Checks that `get(key)` returns `None` when `key` is not cached, and that after calling `put(key, value)` subsequent `get(key)` calls return `value`
2. Set -> Set
- Checks that with duplicated `set(key, value)` calls only the initial call is successful
3. From env var
- Checks that constructing an `InMemoryCache` from an environment variable works.

`AsyncCache` Tests:
1. Get -> Set -> Get
- Same as `Cache` test, but checks both with synchronous and asynchronous execution
2. Set -> Set
- Same as `Cache` test, but checks both with synchronous and asynchronous execution
3. Set -> Set Concurrent
- Checks that of two concurrent `set(key, value)` operations, only one passes

```
cd ~/fbsource/fbcode && buck test mode/opt //caffe2/test/inductor:pcache
```

 {F1981926248}

Rollback Plan:

Differential Revision: D82269762




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

nmacchioni commented on torch/_inductor/pcache.py:
```diff
@@ -0,0 +1,247 @@
+        if env_val is not None:
+            for kv_pair in env_val.split(";"):
+                key, value = kv_pair.split(",")
+                # value is a str repr of bytes, remove b' prefix and ' suffix
```
I added validation and better error messages, closing this out

nmacchioni commented on torch/_inductor/pcache.py:
```diff
@@ -0,0 +1,247 @@
+        raise NotImplementedError
+
+    def get(
+        self: Self, key: Key, executor: ThreadPoolExecutor | None = None
```
I'd rather go forward with ThreadPoolExecutors at this time as to not hold up the other work to figure our asyncio

nmacchioni commented on torch/_inductor/pcache.py:
```diff
@@ -0,0 +1,247 @@
+
+    def get(
+        self: Self, key: Key, executor: ThreadPoolExecutor | None = None
+    ) -> Future[Value | None] | Value | None:
```
I'm not too stuck on it being a single method, I'd be happy to just change it to get_sync and get_async

nmacchioni commented on torch/_inductor/pcache.py:
```diff
@@ -0,0 +1,247 @@
+            The value associated with the key, or None if not found. If an executor is provided,
+            returns a Future representing the result of the asynchronous operation.
+        """
+        if executor is None:
```
Ah ok, so at a high level the code here (Cache and AsyncCache) are just basic building blocks which will be wrapped in a future PR into something called MultiCache. MultiCache will basically just handle get/put calls with multiple backing cache implementations. Another PR will wrap MultiCache into something I'm thinking of calling PCache or maybe just InductorCache. The end goal is that InductorCache provides a function wrapper @InductorCache.record(...) that will be used to replace the existing intra-function recording in pad mm for example. So ultimately I don't expect devs to be interacting with Cache or AsyncCache directly, only through the InductorCache interface

nmacchioni replied:
Let me send you the design doc, it is a touch outdated but it has the general idea of things captured

nmacchioni commented on torch/_inductor/pcache.py:
```diff
@@ -0,0 +1,247 @@
+        try:
+            # "x" mode is exclusive creation, meaning the file will be created
+            # iff the file does not already exist (atomic w/o overwrite)
+            with open(fpath, "xb") as fp:
```
I thought "x" being atomic would ensure that you can't have a partially-written file, but adding a FileLock is so easy I'm just going to take your suggestion there

Skylion007 commented on torch/_inductor/pcache.py:
```diff
@@ -0,0 +1,317 @@
+Value = bytes
+
+
+class Cache:
```
Shouldn't this subclass Generic with defaults of the Key and Value TypeAliases? Would make it more flexible if we need to change it later

nmacchioni replied:
Done; I haven't used Generic previously so please double check my understanding. I ended up using TypeVar as Generic does not take TypeAlias

Skylion007 commented on torch/_inductor/pcache.py:
```diff
@@ -0,0 +1,317 @@
+
+
+# Key represents data used to query from cache instances
+Key = str
```
Please type these as typing_extensions.TypeAlias to make it easier for static analysis

Skylion007 commented on torch/_inductor/pcache.py:
```diff
@@ -0,0 +1,317 @@
+        flock = self._flock_from_fpath(fpath)
+        fpath.parent.mkdir(parents=True, exist_ok=True)
+        try:
+            with flock:
+                # "x" mode is exclusive creation, meaning the file will be created
+                # iff the file does not already exist (atomic w/o overwrite)
+                with open(fpath, "xb") as fp:
```
Have both with contexts on one line to save indentation and improve readability that show they are the same scope

nmacchioni replied:
Done

nmacchioni commented:
@Skylion007 @masnesral do you mind taking a final look? cc @eellison

masnesral reviewed (approved):
LG; thanks.

aorenste commented on torch/_inductor/pcache.py:
```diff
@@ -0,0 +1,315 @@
+        """
+        self._cache: dict[str, bytes] = {}
+
+    def get(self: Self, key: str) -> bytes | None:
```
consider `@override` (here and in the later overrides)

nmacchioni replied:
Done, thanks for the suggestion

aorenste commented on torch/_inductor/pcache.py:
```diff
@@ -0,0 +1,315 @@
+        insert(key, value): Insert a value if the key does not already exist.
+    """
+
+    def get(self: Self, key: Key) -> Value | None:
```
Also - by defining this as `Value | None` it means that you can't have a cache that stores a `None` value... maybe that's fine...

aorenste replied:
That other code has:
```
from __future__ import annotations
```
which enables it in py3.9.

aorenste commented on torch/_inductor/pcache.py:
```diff
@@ -0,0 +1,315 @@
+        Returns:
+            bytes | None: The value associated with the key, or None if not found.
+        """
+        return self._cache.get(key, None)
```
nit: `None` is the default

aorenste commented on torch/_inductor/pcache.py:
```diff
@@ -0,0 +1,315 @@
+        Create a new in-memory cache instance from an environment variable.
+
+        The environment variable should contain key-value pairs separated by ';',
+        with each pair formatted as 'key,value'. The value should be a string
+        representation of bytes (e.g., b'...').
```
So the values must be literally in the form `b'abc'`, no double-quotes, no pure ascii? That seems kind of restrictive?

aorenste commented on torch/_inductor/pcache.py:
```diff
@@ -0,0 +1,315 @@
+
+        if env_val is not None:
+            for kv_pair in env_val.split(";"):
+                if kv_pair == "":
```
nit:
```suggestion
                if not kv_pair:
```

nmacchioni replied:
Fixed

aorenste commented on torch/_inductor/pcache.py:
```diff
@@ -0,0 +1,315 @@
+                    # can happen if env_val is an empty string, or ends with ;
+                    continue
+                try:
+                    key, str_value = kv_pair.split(",")
```
```suggestion
                    key, str_value = kv_pair.split(",", 1)
```

aorenste commented on torch/_inductor/pcache.py:
```diff
@@ -0,0 +1,315 @@
+                # remove b' prefix and ' suffix
+                str_value = str_value[2:-1]
+                value = str_value.encode()
+                # duplicatess are ok, so long as the key does not point to two distinct values
```
nit, spelling:
```suggestion
                # duplicates are ok, so long as the key does not point to two distinct values
```

nmacchioni replied:
Fixed, thanks haha that was metamate

nmacchioni replied:
Thanks! Fixed

aorenste commented on torch/_inductor/pcache.py:
```diff
@@ -0,0 +1,315 @@
+                    key, str_value = kv_pair.split(",")
+                except ValueError as err:
+                    raise ValueError(
+                        f"Malformed kv_pair {kv_pair} in env_var {env_var}!"
```
nit
I would report the string as repr and also indicate a little more "why":
```suggestion
                        f"Malformed kv_pair {kv_pair!r} in env_var {env_var} (missing comma)!"
```

nmacchioni replied:
Gotcha, fixed this and the one below

aorenste commented on torch/_inductor/pcache.py:
```diff
@@ -0,0 +1,315 @@
+                # check that str_value is a str repr of bytes
+                if (not str_value.startswith("b'")) or (not str_value.endswith("'")):
+                    raise ValueError(
+                        f"Malformed value {str_value} in kv_pair {kv_pair}!"
```
nit (same as above)
```suggestion
                        f"Malformed value {str_value!r} in kv_pair {kv_pair} (missing byte string notation)!"
```

aorenste commented on torch/_inductor/pcache.py:
```diff
@@ -0,0 +1,315 @@
+                    )
+                # remove b' prefix and ' suffix
+                str_value = str_value[2:-1]
+                value = str_value.encode()
```
This whole byte encoding thing seems weird - like how would I even put in a non-ascii-7 value? If I want something like char 200 what would I put? It's not at all portable and this encode won't do what you expect. Maybe the input should just be ascii only? Otherwise you really need to handle escape sequences.

masnesral replied:
What is the use case for passing via env var anyway? Without knowing how you'll use it, I would have expected it would be more ergonomic to initialize from the contents of a file. Env vars probably have a max length. Would that be a problem?

nmacchioni replied:
> What is the use case for passing via env var anyway? Without knowing how you'll use it, I would have expected it would be more ergonomic to initialize from the contents of a file. Env vars probably have a max length. Would that be a problem?

End goal is to enable determinism in the compiler, for example we'll have the inductor caching mechanism spit out a history of key -> value mappings from one run which you can plug into the next. It hasn't been fully scoped out yet, but I do think having it point to a file has its own consequences (PGs have to add a file to their resources, to get it included with the binary)

nmacchioni replied:
> Instead of stripping the b and quotes manually you could instead use `ast.literal_eval`:
> 
> ```
> >>> ast.literal_eval("b'abc\\x41'")
> b'abcA'
> ```

I think literal_eval has its own problems, for one we'd still have to check that there is a b' prefix and ' suffix otherwise we'd have to check the return type to make sure it is not evaluating another type

nmacchioni commented on torch/_inductor/pcache.py:
```diff
@@ -0,0 +1,315 @@
+        Returns:
+            bool: True if the value was stored successfully, False if the key already exists.
+        """
+        if key in self._cache:
```
@aorenste we don't overwrite on insert, so key in self._cache is correct here

aorenste reviewed (approved):
The rest of my comments were basically nits (although you should consider the ascii/bin thing) so approving.

aorenste commented on torch/_inductor/pcache.py:
```diff
@@ -0,0 +1,319 @@
+from functools import cached_property
+from os import getenv
+from pathlib import Path
+from typing import Generic, override, TypeVar
```
override is py3.12 - so you need to get it from typing_extensions

nmacchioni committed:
[pcache] Cache and AsyncCache implementations (#162777)

Summary:

Implemented caching abstractions: `Cache` and `AsyncCache`.

`Cache` provides an abstraction for defining simple key -> value stores with get and put functionality. We propose using `Cache` for implementations with very low (microseconds) overhead, for example an in-memory cache.

`AsyncCache` provides an abstraction for defining simple key -> value stores with asynchronous get and put functionality. We propose using `AsyncCache` for implementations with medium to high (> millisecond) overhead, for example an on-disk cache.

We provide an initial extension of `Cache` in the form of `InMemoryCache`. `InMemoryCache` provides fast, in-memory caching that can be later used to memoize more expensive cache accesses. `InMemoryCache` also provides a custom constructor `InMemoryCache.from_env_var` that can be used to pre-populate the in-memory cache, which will be helpful for enabling determinism in the future.

We also provides extensions of `AsyncCache`. `OnDiskCache` subclasses `AsyncCache` and serves as a generic on-disk caching implementation with atomic, write-once guarantees. `OnDiskCache` is semi-generic, allowing subclassing to alter the output directory. `InductorOnDiskCache` subclasses `OnDiskCache` to create an Inductor-specific on-disk cache that outputs to Inductor's default caching directory.

Test Plan:
`Cache` Tests:
1. Get -> Set -> Get
- Checks that `get(key)` returns `None` when `key` is not cached, and that after calling `put(key, value)` subsequent `get(key)` calls return `value`
2. Set -> Set
- Checks that with duplicated `set(key, value)` calls only the initial call is successful
3. From env var
- Checks that constructing an `InMemoryCache` from an environment variable works.

`AsyncCache` Tests:
1. Get -> Set -> Get
- Same as `Cache` test, but checks both with synchronous and asynchronous execution
2. Set -> Set
- Same as `Cache` test, but checks both with synchronous and asynchronous execution
3. Set -> Set Concurrent
- Checks that of two concurrent `set(key, value)` operations, only one passes

```
cd ~/fbsource/fbcode && buck test mode/opt //caffe2/test/inductor:pcache
```

 {F1981926248}

Rollback Plan:

Reviewed By: masnesral

Differential Revision: D82269762
